Disambiguate convert(::Type{Vector{Bottom}}, ::Vector{Bottom})

timholy opened the pull request:
This is the type of method that we deliberately exclude from our ambiguity detection, but it can be triggered by [this call](https://github.com/JuliaLang/julia/blob/ee9d28b199be88bafb69feacc887d4a8464d4376/base/dict.jl#L173). See https://github.com/JuliaImages/ImageCore.jl/issues/40.

tkelman commented on test/abstractarray.jl:
```diff
@@ -840,3 +840,9 @@ end
 end
+
+@testset "ImageCore #40" begin
+    Base.convert(::Type{Array{T,n}}, a::Array) where {T<:Number,n} =
```
why is defining a method like this necessary to reproduce the bug? also, where is `dest` supposed to be coming from?

ararslan replied:
If you don't define this method, you get:
```julia
julia> Base.Test.@test isa(similar(Dict(:a=>1, :b=>2.0), Pair{Union{},Union{}}), Dict{Union{},Union{}})
Test Passed
```
Cool beans. But if you do,
```julia
julia> Base.convert(::Type{Array{T,N}}, a::Array) where {T<:Number,N} = copy!(Array{T,N}(size(a)), a)

julia> Base.Test.@test isa(similar(Dict(:a=>1, :b=>2.0), Pair{Union{},Union{}}), Dict{Union{},Union{}})
Unreachable reached at 0x10a56a351

signal (4): Illegal instruction: 4
while loading no file, in expression starting on line 0
similar at ./dict.jl:193
unknown function (ip: 0x10a56a376)
do_call at /Users/alex/Projects/julia/src/interpreter.c:66
eval at /Users/alex/Projects/julia/src/interpreter.c:246
do_call at /Users/alex/Projects/julia/src/interpreter.c:65
eval at /Users/alex/Projects/julia/src/interpreter.c:246
do_call at /Users/alex/Projects/julia/src/interpreter.c:65
eval at /Users/alex/Projects/julia/src/interpreter.c:246
eval_body at /Users/alex/Projects/julia/src/interpreter.c:552
eval_body at /Users/alex/Projects/julia/src/interpreter.c:598
jl_interpret_toplevel_thunk at /Users/alex/Projects/julia/src/interpreter.c:700
jl_toplevel_eval_flex at /Users/alex/Projects/julia/src/toplevel.c:597
jl_toplevel_eval_in at /Users/alex/Projects/julia/src/builtins.c:496
eval at ./boot.jl:236
jlcall_eval_18469 at /Users/alex/Projects/julia/usr/lib/julia/sys.dylib (unknown line)
eval_user_input at ./REPL.jl:66
jlcall_eval_user_input_20569 at /Users/alex/Projects/julia/usr/lib/julia/sys.dylib (unknown line)
macro expansion at ./REPL.jl:97 [inlined]
#1 at ./event.jl:73
unknown function (ip: 0x10a5627df)
jl_apply at /Users/alex/Projects/julia/src/./julia.h:1440 [inlined]
start_task at /Users/alex/Projects/julia/src/task.c:267
Allocations: 1229937 (Pool: 1228803; Big: 1134); GC: 0
[1]    38285 illegal hardware instruction  ~/Projects/julia/julia
```

ararslan replied:
(Also presumably `dest` is supposed to be `Dict(:a=>1, :b=>2.0)` as in the linked issue)

tkelman replied:
is the convert method wrong then or ? if this changes behavior, should this be careful to always stay the last thing that runs in an isolated test, or try to replicate with a test local type?

ararslan commented:
Looks like 32-bit Windows is still getting the error even with the fix: https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.15147/job/w2ypa6339fpugaxa#L2377

timholy commented:
Thanks for jumping in here, @ararslan! (I'm traveling so only have bits of time here and there.) You're exactly right. And yes, I was trying to ensure that the behavior of that function is the same as it would be without that method definition, for precisely the reason that concerns @tkelman.

But the "Illegal instruction" error is weird and unexpected. Happens on master even without the `Union{}` method defined here. I'll report as a separate error.

timholy committed:
Disambiguate convert & showarg methods with eltype Union{}

This is the type of method that we deliberately exclude from our ambiguity detection, but it can be triggered by, e.g., a grow_to! method on Dicts or trying to display `collect(tuple())`.

timholy commented:
To me this seems ready to merge.
